NACT/attention_query_utilities/attention_query.py

```diff
@@ -441,85 +441,89 @@ def set_scanpy_object(self, new_data):
-                for gene in document:
-                    if gene in freq:
-                        freq[gene] += 1
-                    else:
-                        freq[gene] = 1
-            for gene in freq:
-                freq[gene] = np.log(len(dictionaries) / freq[gene])
-            return freq
-
-        tf_idf = {}
-        # calculate gene frequencies
-        print("==> Calculating gene frequencies")
-        tf_vals = top_n_genes.apply(tf, axis=1)
-
-        # calculate inverse document frequencies
-        print("==> Calculating inverse document frequencies")
-        idf_vals = idf(tf_vals)
-
-        # calculate tf-idf values for all genes in ranked_topN_strdf
-        print("==> Calculating tf-idf values for all ranked genes")
-        for document in range(len(tf_vals)):
-            for gene in tf_vals[document]:
-                tf_idf[gene] = tf_vals[document][gene] * idf_vals[gene]
-        return tf_idf
+
+        # Convert the intial inputs to dataframes 
+        topgenes_df = pd.DataFrame.from_dict(top_n_names) # dataframe containing gene names sorted by attention sums
+        attention_sums = pd.DataFrame.from_dict(top_n_df_dict) # dataframe containing attention sums
+
+
+        # Check if initial input is a dataframe
+        if isinstance(topgenes_df, pd.DataFrame):
+            # Convert input dataframe into a series containing the pseudo documents
+            series = pd.Series(topgenes_df.values.tolist())
+
+
+        # Get the total number of pseudo documents
+        N = series.shape[0]
+
+        #### Calculating the tf-idf values ####
+        # Create a new column to store the gene counts
+        gene_counts = series.apply(lambda x: Counter(x))
+        # Create a new column to store the tf values
+        tf = gene_counts.apply(lambda x: {gene: count/sum(x.values()) for gene, count in x.items()})
+        # Create a new column to store the idf values
+        idf = series.apply(lambda x: {gene:np.log(N/len([i for i in series if gene in i])) for gene in set(x)})
+        # Create a new column to store the tf-idf values
+        tf_idf = tf.apply(lambda x: {gene: count*idf[tf[tf==x].index[0]][gene] for gene, count in x.items()})
+
+        # Create a dataframe of the tf-idf values
+        df_tfidf = pd.DataFrame(tf_idf.tolist())
+        # Group the dataframe by gene and calculate the sum of the tf-idf scores
+        df_tfidf = df_tfidf.sum()
+
+        # Normalizing the values to make sure that common genes are closer to 0
+        df_tfidf = (1 - df_tfidf/np.log(N))
+
+
+        #### tf-idf normalization of attention sums ####
+        # Get tf-idf genes to index the attention sum dataframe
+        tfidf_genes = df_tfidf.index.tolist()
+
+        # Subset the dataframe to only contain the top genes
+        at_sub = attention_sums.loc[tfidf_genes, :]
+
+        # Perform element-wise muliplication
+        at_norm = attention_sums.mul(df_tfidf, axis = 0) # element wise multiplication
```

Comment:
**R**: `at_norm` is not descriptive enough (see [SG](https://google.github.io/styleguide/pyguide.html#316-naming))